overwatch-rs/src/services/relay.rs

```diff
@@ -105,13 +112,16 @@ impl<M> InboundRelay<M> {
     }
 }
 
-impl<M> OutboundRelay<M> {
+impl<M: Send + 'static> OutboundRelay<M> {
     /// Send a message to the relay connection
-    pub async fn send(&self, message: M) -> Result<(), (RelayError, M)> {
-        self.sender
-            .send(message)
-            .await
-            .map_err(|e| (RelayError::Send, e.0))
+    pub async fn send(&mut self, message: M) -> Result<(), (RelayError, M)> {
```

Comment:
I had to change the signature to use a `&mut self`. It should not matter iirc, as we only use one hook per end. But lets see if this cause a problem. The alternative is having a method that consumes the object and returns a Sink (as I had in an earlier commit), but it ends up in with the same issue as having a `&mut`, so it should be ok.

Reply:
How about we use `async-channel` here? `Sender` and `Receiver` in `async-channel` are interior mutability. I think we can make overwatch support async runtime agnostic.

Reply:
It shouldn't be too difficult to do that, and I had that in my mind already. Feel like giving it a shot @al8n?
IIRC, we would only need to change the channels and the calls to `tokio::spawn` for something that abstracts it (I remember there is something like `executor::spawn` in some crate 🤔 ).

We can keep it like this for now, and make a full refactor of it.

Reply:
For `spawn` stuff, we can do something like 
```rust
#[inline]
pub fn foo<SP, R>(
        spawner: SP,
) -> Result<(), Error>
where
        SP: Fn(BoxFuture<'static, ()>) -> R + Send + Sync + 'static + Copy, {
     ...
}

foo(tokio::spwan);
foo(async_std::thread::spawn)
```

Reply:
Also would like to avoid the possibility of having multiple different runtimes.
IMO as long as we don't make strong assumptions about the runtime (e.g. there is a separate thread for events, ...) it's fine to hardcode it. 
Either we need to support different runtimes in different deployments, or simply changing to another one in case tokio it's not maintained anymore or is not performing at our expectations should be relatively simple.

Reply:
> I had to change the signature to use a &mut self

A few ideas to avoid this:
* Can we make the conversion from `Sender` to `PollSender` on the fly only when we need the `Sink` functionalities?
* Keep both a `Sender` and a `PollSender` on the same `OutboundRelay`
* Do not implement `Sink` directly (which is the same strategy used by tokio but add a method like `into_sink` that returns a type that implements `Sink<Item = T>`)

Reply:
>> Do not implement Sink directly (which is the same strategy used by tokio but add a method like into_sink that returns a type that implements Sink<Item = T>)

Yeah, First iteration was like this and I think it was simpler and remove the problems. Gonna bring it back.